### ISSUE TITLE
deprecated case class warns at its definition

### ISSUE THREAD

**Issue by bishabosha** [lampepfl/dotty#11022]
## Minimized code

```Scala
@deprecated("no CaseClass", "0.1") case class CaseClass(rgb: Int)
```

## Output

```scala
-- Deprecation Warning: CaseClass.scala:1:0 ----
21 |@deprecated("no CaseClass", "0.1") case class CaseClass(rgb: Int)
   |^
   |class CaseClass is deprecated since 0.1: no CaseClass
-- Deprecation Warning: CaseClass.scala:1:46 ---
21 |@deprecated("no CaseClass", "0.1") case class CaseClass(rgb: Int)
   |                                              ^
   |                   class CaseClass is deprecated since 0.1: no CaseClass
```

## Expectation

no warning as in Scala 2.13.4.

**Comment by gregbeech** [comment]
This is also an issue for deprecated implicit classes. I assume it's the same basic issue so I'll just add this as a comment rather than opening a new issue.

# Minimized Code

```scala
object repro {
  @deprecated("no ImplicitClass", "0.1") implicit class ImplicitClass(val a: Any)
}
```

# Output

```
[warn] 2 |  @deprecated("no ImplicitClass", "0.1") implicit class ImplicitClass(val a: Any)
[warn]   |  ^
[warn]   |class ImplicitClass in object repro is deprecated since 0.1: no ImplicitClass
```

# Expectation

No warning, as in Scala 2.13.6

**Comment by som-snytt** [comment]
I was going to address it after https://github.com/scala/scala/pull/10071 which ... how did I put it ...

> Instead of juggling annotations, the deprecation loophole is expanded to look more like the tax code.

**Comment by He-Pin** [comment]
Any update about this @som-snytt ,thanks.

**Comment by som-snytt** [comment]
@He-Pin by coincidence, yesterday I deleted the old branch from a year ago, but got distracted by warning on deprecated constant-folded ops. I'll return to this right after that. The interesting lesson from that other fix is that because of structural differences in the pipeline, fixes don't necessarily forward-port directly.

**Comment by jan-pieter** [comment]
When adding a companion object with a deprecated annotation, the warning disappears:
```scala
@deprecated("no CaseClass", "0.1")
case class CaseClass(rgb: Int)

@deprecated("no CaseClass", "0.1")
object CaseClass
```

**Comment by som-snytt** [comment]
The draft PR is the Scala 2 fix, which specifies suppression by deprecated companion of enclosing element, but the fix is less fussy than Scala 2, at least as a first attempt.

Previously excluded was the linked issue https://github.com/lampepfl/dotty/issues/12706

**Comment by som-snytt** [comment]
I haven't had a chance to return to this issue (both Scala 2 and 3) to evaluate whether my proposal is a bad idea. (That must happen before 2.13.11.) Thanks for the bump.

**Comment by som-snytt** [comment]
Um, I had an idea but maybe don't go down that alley. Thanks, scala-center-bot!

**Comment by SethTisue** [comment]
> As there now is a work-in-progress pull request (https://github.com/lampepfl/dotty/issues/12706) fixing this issue

I guess you mean #16918

**Comment by mbovel** [comment]
> I guess you mean https://github.com/lampepfl/dotty/pull/16918

Indeed! I fixed it in my original comment.

**Comment by SethTisue** [comment]
The Scala 2 PR for this, scala/scala#10071, was merged a while back.

@som-snytt thanks for your helpful remarks above, last week. 

The approach taken in the Scala 2 PR (namely "only exclude the case ctor call") seems satisfactory to me, but I guess we should begin our session by seeing whether we agree on that.

**Comment by SethTisue** [comment]
@nmcb @jan-pieter wdyt about this as the possible culprit?

```scala
  override def transformNew(tree: New)(using Context): New = {
    checkUndesiredProperties(tree.tpe.typeSymbol, tree.srcPos)
    tree
  }
```

`New` is a constructor call, but note that the symbol passed is `tree.tpe.typeSymbol`

**Comment by jan-pieter** [comment]
Both the `transformNew` and `transformTypeTree` entrypoints were triggering the warning. Do you think we need to only apply the exemption defined [here](https://github.com/lampepfl/dotty/pull/17165/files#diff-20d26ac73075339cc92ba7fcec42405d2b78c8f852ef8f38ab3be4bafe430d07R65) when coming from those entrypoints?
